### ISSUE TITLE
CSS find-in-page highlight pseudos

### ISSUE THREAD

**Issue by jihyerish** [w3ctag/design-reviews#1120]
こんにちは TAG-さん!

I'm requesting a TAG review of the new highlight pseudo-element, **::search-text**

This pseudo-element exposes find-in-page search result styling to authors, including selection and spelling errors. It allows authors to change the foreground and background colors or add text decorations, which can be especially useful if the UA defaults are insufficiently contrasted with the page colors or otherwise unsuitable.

### Specification

https://drafts.csswg.org/css-pseudo-4/#selectordef-search-text

### Explainer

https://github.com/Igalia/explainers/blob/main/css/find-in-page/README.md

### Links

- An introduction to the feature, aimed at unfamiliar audiences: https://github.com/Igalia/explainers/blob/main/css/find-in-page/README.md
- A description of the problems that end-users were facing before this proposal: https://stackoverflow.com/questions/50309703/css-for-browsers-find-in-page
- Examples of how to use the proposal to solve the end-users' problems: https://github.com/Igalia/explainers/blob/main/css/find-in-page/README.md#example
- Security and Privacy self-review: https://docs.google.com/document/d/1qcq2kQRXgAP67wk9aIfECAq4C_JcJN5Zw1Tx7jDnFvI/edit?tab=t.0#heading=h.kr0x0i2en8w3
- Web Platform Tests:  third_party/blink/web_tests/wpt_internal/css/css-pseudo/search-text-*

### The specification

- [x] Follows the [Web Platform Design Principles](https://www.w3.org/TR/design-principles/).
- [x] Includes Security and Privacy Considerations sections based on answers to the [Security/Privacy Questionnaire](https://www.w3.org/TR/security-privacy-questionnaire/).

### Where and by whom is the work is being done?

- GitHub repo: https://github.com/Igalia/explainers/tree/main/css/find-in-page
- Primary contacts:
  - Delan Azabani (@delan), Igalia
  - Jihye Hong (@jihyerish), Igalia
  - Stephen Chenney (@schenney-chromium), Igalia
- Organization/project driving the specification: Igalia
- This work is being funded by: Bloomberg and Igalia
- Primary standards group developing this feature: CSSWG
- Incubation and standards groups that have discussed the design:
  - https://github.com/w3c/csswg-drafts/issues/3812 This is the initial idea of this feature. After several discussions, the syntax has been changed

### Feedback so far

- Multi-stakeholder feedback:
  - Chromium comments: Positive, https://chromestatus.com/feature/5195073796177920
  - Mozilla comments: https://github.com/mozilla/standards-positions/issues/1103
  - WebKit comments: https://github.com/WebKit/standards-positions/issues/421
- Status/issue trackers for implementations: [Chromium](https://issues.chromium.org/issues/339298411)

---

Track conversations at https://tag-github-bot.w3.org/gh/w3ctag/design-reviews/1120

**Comment by matatk** [comment]
Hi @jihyerish, the TAG discussed it at a Breakout today, and found that:
- The use case is strong, and the proposed feature solves it nicely while being consistent with the existing highlight pseudos
- However, we are concerned that the feature is not implementable on Safari, and therefore, will break the interoperability of the Web

Have you explored alternatives that can mitigate the interoperability issue? A good starting point is https://www.w3.org/TR/design-principles/#implementability

Thank you!

**Comment by schenney-chromium** [comment]
Thanks for the positive feedback on the use cases.

Regarding Safari not implementing, two things come to mind:
* The reason Safari will not implement is that they have a vendor-specific solution to the contrast problem implemented at the browser UI level. In particular, Safari find-in-page puts an overlay over the page to ensure contrast with the search UI. This overlay both addresses the use cases and prevents page-level control.
* The missing feature will be detectable in Safari via @supports.

I think the Safari UI would be an obvious flag to developers that custom search markers do not work, and that UI also prevents users from suffering significantly if a page fails to note the lack of compatibility.

**Comment by matatk** [comment]
Thanks for your reply @schenney-chromium. We discussed this proposal again today, and are still supportive of the use case (improving the in-page highlight experience and accessibility). We have an alternative approach suggestion for you to consider.

The explainer talks about the behaviours of different browsers. However, did you consider the alternative approach of _not_ implementing this, and instead setting the expectation the UAs be responsible for choosing highlight colors based on the active background color and user preferences so that they provide appropriate contrast (similar to Safari's more involved approach)?

Taking this alternative approach would not afford content authors the same level of control (and responsibility) that the highlight pseudos approach provides. It would however have the effect of making the 'find in page' feature consistent within a given UA, and it would avoid the possibility that authors fail to provide appropriate colors. (Given that some users may prefer lower, or higher, contrast, and dark mode may be in effect, providing appropriate colors is non-trivial work for the content author.)

In addition, have you considered possible abuse scenarios? I.e. under the proposed approach, authors could set colors that effectively disable the find-in-page feature.

**Comment by bramus** [comment]
> In addition, have you considered possible abuse scenarios? I.e. under the proposed approach, authors could set colors that effectively disable the find-in-page feature.

I don’t think this is a strong argument, as the same thing can be said about _any_ CSS feature.  E.g. What’s preventing an author from setting a text’s color to white on a white background? What’s preventing an author from giving an element an opacity of 0? What’s preventing an author from setting a ridiculously small font-size? What’s preventing authors from removing underlines from links? …

**Comment by bramus** [comment]
@matatk Thanks for clarifying. I read your initial comment more as an argument against the feature, whereas you want to make sure that this question was covered as part of the design process.

**Comment by matatk** [comment]
Hi again all. We have just discussed this again in a TAG call, and we see that [the blink-dev discussion](https://groups.google.com/u/0/a/chromium.org/g/blink-dev/c/U-6tIuuGtgo/m/TgCkWVWoAQAJ) is progressing. As is being discussed there, it would be great to have more input from developers as to their needs (if you have any research on this, please do also include it in the explainer).

We were wondering in our call: could you use something like `:has(::search-text)` on a table row to cause it to be highlighted if the row contains a match somewhere? Are there other things that you could _only_ do if this was shipped (as opposed to having the UA control the colors)?

We're also wondering if you are planning to pause on this whilst the blink-dev discussion is in progress (looking at the thread, it seems so), or whether you're intending to ship this proposal soon?

I should've labelled this issue appropriately to indicate that we await the explainer updates discussed above, so I'll do that now. We look forward to following the blink-dev discussion, and hearing more here as you have it. Thanks again.

P.S. If it's of interest, the URL for [our minutes for this week](https://github.com/w3ctag/meetings/blob/gh-pages/2025/telcons/09-08-minutes.md) should work at the end of this week (time zone-dependent).

**Comment by schenney-chromium** [comment]
Sorry for letting this drop for so long.

Regarding the `:has(::search-text)` question, this will only tell you whether the style would apply _if_ the text is highlighted. It doesn't tell you whether that style is active of not. This is the same as `getComputedStyle`, which will return the style that would apply but not whether that style is active. The behavior is designed to protect user privacy and matches that for `spelling-error` and `::grammar-error` that have similar concerns.

**Comment by schenney-chromium** [comment]
I've also done some more searching for developers or users asking for the ability to modify the search results appearance. Through the user style sheet or a browser extensions individual users could make use of `::search-text` to increase contrast for the find results even if sites do not. This would be useful for users who desire higher contrast or other accommodations.

On the original CSS WG issue someone wants this to avoid conflicts with the ECMAScript spec highlights. https://github.com/w3c/csswg-drafts/issues/3812#issuecomment-1703272181

In a discussion about find-in-page and accessibility, a "remaining challenge" is "Styling Search Matches" https://schepp.dev/posts/rethinking-find-in-page-accessibility-making-hidden-text-work-for-everyone/#remaining-challenges

A Mozilla user asks for the feature here: https://connect.mozilla.org/t5/ideas/increased-visibility-for-current-quot-find-in-page-quot-match/idi-p/26500

**Comment by matatk** [comment]
Thanks @schenney-chromium for your updates on this thread. Sorry for my slow reply in return.

There's a key question we should discuss: is finding text on the page a feature of a _web page_, or a feature of the UA? Should users have an expectation that the feature will work in the same way across sites? If a feature is inconsistent (even in appearance) it could confuse users, or lead them to believe that only some pages provide it.

A much more extreme example of this would be the 'toolbar' widgets that can be added to pages, to allow the user to customise the page's appearance (font size, colour, etc.) These provide useful features, but exposing them on a per-page basis can lead users to believe that only some pages support them. Whilst it's an extreme example, it does show that consistency and discoverability are important factors.

Allowing authors to be able to customise the way that find-in-page is presented seems like a reasonable idea. Ensuring that the default presentation of this feature is accessible is critical. Personally, I'd really want UAs to offer an option to let me override any author-specified colours here - at least providing users the choice to always use the UA's known-accessible appearance. (I have a vision impairment, but am trying to be as general as possible.)

In the case that authors _do_ provide their own colours, we should at least advise the authors that this means they'll need to ensure sufficient contrast in...

* light and dark display modes

* situations where the design of the site (e.g. foreground/background) change - this also invites the question as to whether the colours would need to remain consistent, or would react to the site design (each of those possibilities have different potential costs).

We have discussed the way that Safari renders the find-in-page UI in this thread. That rendering - in which the page context _except_ for matches is dimmed - is part of how the UA guarantees sufficient contrast. Whilst we would not seek to tell UAs how to do their UX, it's worth noting that this rendering isn't possible for authors to achieve within the current proposal, because the parts of the page that aren't matches cannot have their appearance modified.

In summary:

* Is find-in-page a _page_ or a _UA_ feature?

* Have you considered whether the UA might render the results in a more profound way than the spec supports?

**Comment by schenney-chromium** [comment]
> Regarding the `:has(::search-text)` question, this will only tell you whether the style would apply _if_ the text is highlighted. It doesn't tell you whether that style is active of not. This is the same as `getComputedStyle`, which will return the style that would apply but not whether that style is active. The behavior is designed to protect user privacy and matches that for `spelling-error` and `::grammar-error` that have similar concerns.

I was wrong about this. Pseudo-element selectors are not allowed in `:has` so  `:has(::search-text)` will never match.

**Comment by schenney-chromium** [comment]
In reply to https://github.com/w3ctag/design-reviews/issues/1120#issuecomment-3790677450

The MDN entry for `::search-text` has been written and explicitly discusses contrast issues with links and advice on how to test your site. I would expect that has some impact on developers hoping to use the feature.

Find-in-page is a UA feature that some sites aim to circumvent. Some UAs provide better interfaces for presenting the results, and one could certainly push for browser vendors to improve their UI. At some point I got the colors for Chrome's find-in-page ticks changed to improve contrast, but obviously more can be done as demonstrated by Safari (though I personally found their UI somewhat confusing).

Having said that, many sites choose to capture Ctrl-F and implement their own search functionality. I believe that most users would not be able to explain the distinction between the page's search box and the browser UI search box, particularly since only one is displayed if you hit Ctrl-F in most cases (or both present would be even more confusing). Pages may choose this route to improve contrast or the user's search experience.

So I guess it's both a page and UA feature.

Other UA features may also be styled by CSS: spell check result, grammar results and text-fragment URL matches.
